### ISSUE TITLE
Disk space efficient package manager for Python

### ISSUE THREAD

**Issue by ChillarAnand** [pypa/packaging-problems#328]
When I run `$ du -hs *` in `~/.virtualenvs`, it shows that the size of most of them is over 400MB. The total space consumed by `virtualenvs` is over 6GB. Also, there are a lot of packages which are common across all the environments and there are few packages which are different. 

Is there a better way to organize the dependencies so that disk space can be reduced? 

In node.js ecosystem, there is [pnpm](https://github.com/pnpm/pnpm) which solves this problem by having a global store and then linking it wherever it is required. Is there any such package manager for Python packages?

**Comment by KOLANICH** [comment]
>Disk space efficient package manager

vennvs **IS NOT** a **package manager** and mustn't be used as such. Use **pip** to install packages systemwide or userwide.

**Comment by uranusjr** [comment]
> In node.js ecosystem, there is pnpm which solves this problem by having a global store and then linking it wherever it is required. Is there any such package manager for Python packages?

There is not AFAIK, but the same idea would be useful here. To do this manually, you would do something like (code illustrative only)

```
# Install package to a prefix.
# PREFIX is a prefix to store all packages in.
# PACKAGE_ID should be some kind of scheme to recognise different
# versions of a package (so you can have different package versions, and avoid
# mixing Numpy built for Python 3.8 with Numpy built for 3.7).
python3.8 -m pip install --target PREFIX/numpy/PACKAGE_ID numpy

# Write an .pth file to include the directory in sys.path.
# Delete this to “uninstall” the package.
echo 'PREFIX/numpy/PACKAGE_ID' > PATH_TO_ENV_SITE_PACKAGES/numpy.pth
```

I can imagine it being not too difficult to cobble together some simple solution with some extra metadata to clean up a package when no enviroments are referring to it anymore. Give it a try! (And please share on distutils-sig or discuss.python.org if you come up with something.)

**Comment by ChillarAnand** [comment]
@KOLANICH I agree. virtualenv is not a package manager. There are several projects which have `requirements.txt` file and I use pip to install these packages in virtualenv which results in increased disk space. 

@uranusjr Thanks for the pointers. I will try to write a small tool and see how it goes.

**Comment by ncoghlan** [comment]
On Mon., 24 Feb. 2020, 2:20 am Chillar Anand, <notifications@github.com>
wrote:

> @KOLANICH <https://github.com/KOLANICH> I agree. virtualenv is not a
> package manager. There are several projects which have requirements.txt
> file and I use pip to install these packages in virtualenv which results in
> increased disk space.
>
> @uranusjr <https://github.com/uranusjr> Thanks for the pointers. I will
> try to write a small tool and see how it goes.
>

If you're looking for an example of an existing venv management tool that
offers this feature, you may want to take a look at "pew add":
https://github.com/berdario/pew/blob/master/README.md#add

pew also has several other nice features for managing complex multi-venv
setups (like "pew inall <command to run in each managed venv>"), so you may
find it to be worth a look in general (rather than just as inspiration)

Cheers,
Nick.

**Comment by msarahan** [comment]
> virtualenv is not a package manager

A better term is perhaps "environment manager" along with the concept of a central package cache that may be shared efficiently among more than one environment.  Conda is a good example here of something that has the central package cache that is then linked into environments.  It is very different from most other python environment managers in that it includes python itself as something that it manages.  It is much more like a system-level package manager that can manage environments, but which is separate from the system.  Many people don't like that it doesn't use a system Python installation, but it does help avoid dumb things happening to the system python when someone uses "sudo" without knowing what they're doing.

People can use pip to install packages from PyPI into conda environments, but it usually works better if you only use conda to manage conda environments.  This is because the metadata for either ecosystem is not completely aligned, and stuff can get accidentally broken.

Anyway, to each their own.  Good luck.  I hope you don't end up needing to create a new tool here.

**Comment by KOLANICH** [comment]
> It is much more like a system-level package manager that can manage environments

Yes, I usually install stuff like openssl and sqlite using it on Windows. The main drawbacks of conda is that it is damn slow, a bit glitchy, very fragile in terms of dependencies and stores a cache of old shit (it can be cleaned, but conda is damn slow, so cleaning is also damn slow). One shouldn't use smt solvers when creating package managers.

> Many people don't like that it doesn't use a system Python installation, but it does help avoid dumb things happening to the system python when someone uses "sudo" without knowing what they're doing.

It's better to use sudo with knowing what it does rather than use conda on *nixes IMHO.

>People can use pip to install packages from PyPI into conda environments, but it usually works better if you only use conda to manage conda environments. This is because the metadata for either ecosystem is not completely aligned, and stuff can get accidentally broken.

Conda detects pip-installed packages. In fact the main reason to use conda is because there are some packages in conda that are not officially distributed prebuilt for Windows. Anaconda distributes them prebuilt.

**Comment by gaborbernat** [comment]
> If you're looking for an example of an existing venv management tool that offers this feature, you may want to take a look at "pew add": https://github.com/berdario/pew/blob/master/README.md#add

If one does not want to learn a new tool one could use virtualenv 20 plugin system to automatically maintain some local database entry upon environment creation, and similarly, shim the ``pip`` command within it to also track installed packages, and perhaps instead install them via the ``.pth files``. This would have the advantage of things would work for all tools, such as pipenv, tox, nox, etc... basically anything using virtualenv.

**Comment by ChillarAnand** [comment]
> If you're looking for an example of an existing venv management tool that
offers this feature, you may want to take a look at "pew add":

@ncoghlan Yes, I am looking for existing tools that offer this feaure. pew seems to be a useful tool. However, in this case it won't work as it can add entire `site-packages` of an env not a single package. 

For example, consider 2 projects. project1 with django==2.0 & numpy==1.18 as dependencies and project2 with django==3.0 &  numpy==1.18 as dependencies. Now, in project2 running `pew add project1_site_packages` will break django dependency?

@msarahan I didn't know conda had this feature. I will try it out and see. 

@gaborbernat I am not sure if virtualenv provides hooks to shim pip. But if it works, this is a better plan than writing a new tool.

**Comment by gaborbernat** [comment]
It does provide you with the mechanism to define a custom seed mechanism... basically allowing you to provide an alternative pip version 👍 https://virtualenv.pypa.io/en/latest/extend.html#seed-mechanism

**Comment by ncoghlan** [comment]
On Mon., 24 Feb. 2020, 8:32 pm ChillarAnand, <notifications@github.com>
wrote:

> If you're looking for an example of an existing venv management tool that
> offers this feature, you may want to take a look at "pew add":
>
> @ncoghlan <https://github.com/ncoghlan> Yes, I am looking for existing
> tools that offer this feaure. pew seems to be a useful tool. However, in
> this case it won't work as it can add entire site-packages of an env not
> a single package.
>
> For example, consider 2 projects. project1 with django==2.0 & numpy==1.18
> as dependencies and project2 with django==3.0 & numpy==1.18 as
> dependencies. Now, in project2 running pew add project1_site_packages
> will break django dependency?
>
Not necessarily, as the shared directory is appended to the path, so it's
possible to shadow dependencies without uninstalling them first, similar to
what happens in a virtual environment that can see the system site packages
directory.

To be most effective, the shared directory model is best combined with
pip's "constraints file" feature, so projects use the same versions of
dependencies, but also don't install packages they don't need.

It's also possible to get more exotic with your shared directory layouts
(e.g. breaking them up by project name and major version), but then you
have a more complex cache management problem to solve.

Cheers,
Nick.

>

**Comment by ChillarAnand** [comment]
@uranusjr suggestion is working well. I will experiment with the script for some time and see how it goes. 

> Not necessarily, as the shared directory is appended to the path, so it's
possible to shadow dependencies without uninstalling them first, similar to
what happens in a virtual environment that can see the system site packages
directory.

@ncoghlan  I didn't get this part. 

I tried the following with 2 virtual environments with the 2 projects as mentioned earlier. This seems to break the dependency? 

```
(d30) # curatech @ anands-MacBook-Pro-2 in ~ [19:48:54] 
$ python -c 'import django; print(django.VERSION)'
(3, 0, 0, 'final', 1)

(d30) # curatech @ anands-MacBook-Pro-2 in ~ [19:48:56] 
$ pew add ~/.virtualenvs/d20/lib/python3.7/site-packages      

(d30) # curatech @ anands-MacBook-Pro-2 in ~ [19:49:03] 
$ python -c 'import django; print(django.VERSION)'      
(2, 0, 0, 'final', 0)
```

I came across some NumPy issue where a single package size can also be reduced. Not sure what is happening internally but this also seems to reduce the space dramatically. 

```
$ pip install numpy

$ du -hs ~/.virtualenvs/all3/lib/python3.7/site-packages/numpy/ 
 85M	/Users/avilpage/.virtualenvs/all3/lib/python3.7/site-packages/numpy/

$ CLFAGS='-g0 -Wl -I/usr/include:/usr/local/include -L/usr/lib:/usr/local/lib' pip install numpy --no-cache-dir --global-option=build_ext

$ du -hs ~/.virtualenvs/py37/lib/python3.7/site-packages/numpy/
 23M	/Users/avilpage/.virtualenvs/py37/lib/python3.7/site-packages/numpy/
```

**Comment by ncoghlan** [comment]
@ChillarAnand Try running "python -m site" to see what Python import paths are being added, and their relative order.

**Comment by ChillarAnand** [comment]
```
$ python -m site                                  
sys.path = [
    '/Users/avilpage/projects/library/',
    '/Users/avilpage/.virtualenvs/d20/lib/python3.7/site-packages',
    '/Users/avilpage/.virtualenvs/d30/lib/python37.zip',
    '/Users/avilpage/.virtualenvs/d30/lib/python3.7',
    '/Users/avilpage/.virtualenvs/d30/lib/python3.7/lib-dynload',
    '/usr/local/Cellar/python/3.7.5/Frameworks/Python.framework/Versions/3.7/lib/python3.7',
    '/Users/avilpage/.virtualenvs/d30/lib/python3.7/site-packages',
]
USER_BASE: '/Users/avilpage/.local' (exists)
USER_SITE: '/Users/avilpage/.local/lib/python3.7/site-packages' (doesn't exist)
ENABLE_USER_SITE: False
```

Looks like pew added directory(`/Users/avilpage/.virtualenvs/d20/lib/python3.7/site-packages`) is at top instead of bottom? @ncoghlan

**Comment by ChillarAnand** [comment]
It looks like pew placing this directory at top instead of bottom. 

```
$ cat ~/.virtualenvs/d30/lib/python3.7/site-packages/_virtualenv_path_extensions.pth 
import sys; sys.__plen = len(sys.path)
/Users/avilpage/.virtualenvs/d20/lib/python3.7/site-packages
import sys; new=sys.path[sys.__plen:]; del sys.path[sys.__plen:]; p=getattr(sys,'__egginsert',0); sys.path[p:p]=new; sys.__egginsert = p+len(new)
```
Commenting the last line solves the problem.

**Comment by ncoghlan** [comment]
My apologies, I had no idea `pew add` used the pth file code execution loophole to create an override directory rather than simply creating a normal .pth file that appends a directory to sys.path.

So, yeah, for your use case, you definitely don't want the override behavior - just a plain old vanilla pth file that lets you chain multiple venvs together, with the app specific ones taking priority over the shared ones.

**Comment by ChillarAnand** [comment]
`pew add`(or sharing site-packages with .pth files) works well to share a set of common packages with other environments. 

Lets assume numpy==1.18.1 is present in a common environment and is shared with other environments. If numpy==1.17 is required for 2 other projects, then we still end up with installing numpy==1.17 twice instead of only once. Is there a better way of handling this? @ncoghlan
